### PR TITLE
Feature/GAL-1183

### PR DESCRIPTION
**Pull request by MichaelW-NHS**
## Description

Event Notification: Lambdas - Send notify message

https://nhsd-jira.digital.nhs.uk/browse/GAL-1183
https://nhsd-jira.digital.nhs.uk/browse/GAL-1149
https://nhsd-jira.digital.nhs.uk/browse/GAL-1198

sendSingleNotifyMessage lambda created in order to pick up messages from the enriched queue and sending them one at a time to NHS Notify. The result is then stored in NotifySendMessageStatus table.

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+  for (let record of records) {
+    try {
+      const messageBody = JSON.parse(record.body);
+      const messageReferenceId = await generateMessageReference();
+      const messageSentAt = new Date().toISOString();
+      try {
+        const {responseObject, numberOfAttempts} = await sendSingleMessage(messageBody, accessToken, messageReferenceId, process.env.MESSAGES_ENDPOINT_URL, process.env.INITIAL_RETRY_DELAY, process.env.MAX_RETRIES);
+        const responseBody = responseObject.data;
+        console.log(`Request to NHS Notify for ${messageBody.participantId} successful  - MessageId: ${responseBody.id}`);
```
Comment (chenglawson1) [inline terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js:48]: Might be useful to include the routing plan id in the log message so it's easy to see which message has been sent.

File: terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        await putSuccessResponseIntoTable(messageBody, messageSentAt, numberOfAttempts, responseBody, messageReferenceId, notifySendMessageStatusTable, dynamodbClient);
+        recordsSuccessfullySent++;
+      } catch(error) {
+        if(error.status && error.details) {
+          console.error(`Error: Request to NHS Notify for participant ${messageBody.participantId} failed - Status code: ${error.status} and Details: ${error.details}`)
```
Comment (chenglawson1) [inline terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js:53]: Might be useful to include the routing plan id.

File: terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  let messageSentAt = new Date().toISOString();
+
+  const response = await customFetch(messagesEndpoint, {
+    method: 'POST',
+    retryDelay: (attempt) => initialRetryDelay * Math.pow(2, attempt),
+    retryOn: (attempt, error, response) => {
+      messageSentAt = new Date().toISOString();
+      if (statusCodesToRetryOn.includes(response.status) && attempt < maxRetries) {
+        numberOfAttempts++;
+        console.error(`NHS Notify request failed for participant ${participantId} with status code ${response.status}`);
```
Comment (chenglawson1) [inline terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js:223]: Please add prefix to message "Error:".

File: terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+  try {
+    const response = await client.send(
+      new GetSecretValueCommand({
+        SecretId: secretName,
+      })
+    );
+    console.log(`Retrieved value successfully ${secretName}`);
+    return response.SecretString;
+  } catch (error) {
+    console.log(`Failed: ${error}`);
```
Comment (chenglawson1) [inline terraform/src/sendSingleNotifyMessageLambda/sendSingleNotifyMessageLambda.js:275]: Please change prefix to "Error:" and console.error.

**Review by chenglawson1 — APPROVED** [review]

**Review by Serena-NHS — APPROVED** [review]

**Pull request merged**